gensim/models/experimental/custom_callbacks.py

```diff
@@ -0,0 +1,72 @@
+import logging
+try:
+    from keras.callbacks import Callback
+    KERAS_AVAILABLE = True
+except ImportError:
+    KERAS_AVAILABLE = False
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(
+    format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO)
+
+
+class ValidationCallback(Callback):
+    """Callback for providing validation metrics on the model trained so far"""
+    def __init__(self, test_data):
+        """
+        Parameters
+        ----------
+        test_data : dict
+            A dictionary which holds the validation data. It consists of the following keys:
+                - "X1" : numpy array
+                    The queries as a numpy array of shape (n_samples, text_maxlen)
+                - "X2" : numpy array
+                    The candidate docs as a numpy array of shape (n_samples, text_maxlen)
+                - "y" : list of int
+                    It is the labels for each of the query-doc pairs as a 1 or 0 with shape (n_samples,)
+                    where 1 : doc is relevant to query, 0 : doc is not relevant to query
+                - "doc_lengths" : list of int
+                    It contains the length of each document group. I.e., the number of queries
+                    which represent one topic. It is needed for calculating the metrics.
+
+        """
+
+        if not KERAS_AVAILABLE:
+            raise ImportError("Please install Keras to use this class")
+
+        # Check if all test_data is a dicitonary with all the right keys
+        try:
+            # If an empty dict is passed
+            if len(test_data.keys()) == 0:
+                raise ValueError(
+                      "test_data dictionary is empty. It doesn't have the keys: 'X1', 'X2', 'y', 'doc_lengths'"
+                    )
+            for key in test_data.keys():
+                if key not in ['X1', 'X2', 'y', 'doc_lengths']:
+                    raise ValueError("test_data dictionary doesn't have the  keys: 'X1', 'X2', 'y', 'doc_lengths'")
```

Comment:
incorrect check: if `test_data.keys()` contains needed keys + some additional key - this will fail